# Case


* [1. The needed steps](#needs)
    * [1.1. Import Libraries](#import)
    * [1.2. Import functions](#import2)
    * [1.3. Import dataset](#import3)
* [2. Data Exploration](#exploration)
    * [2.1. Summary statistics](#missing)  
* [3. Questions](#questions)
    * [3.1. How many containers are exported and imported from the port?](#q1)
    * [3.2. How many containers are exported and imported per shipsystem (SSY)? SSYs are the
different ocean services which are connected to voyages. In the data set, the more
generalized term MOT_Service is used (MOT = Mode of Transport)](#q2)
    * [3.3. Display the IoT on a map by highlighting the points and by displaying the routes per move?](#q3)
    * [3.4. What is the maximum distance a container traveled from the hinterland to a port (and vice
versa)?](#q4)
    * [3.5. Calculate the travel time of the containers and give an insight on the typical travel times of
the containers.](#q5)
    * [3.6. General: Share your observations on the data?](#q6)

<hr>
<a class="anchor" id="needs">
    
# 1. The needed steps
    
</a>


<a class="anchor" id="import">

## 1.1. Import Libraries

</a>


In [1]:
import pandas as pd
import numpy as np
import folium
import random

import sys
sys.path.insert(0, '/Users/maceli/useful_functions')

from functions import *

In [2]:
add(1,4)

5


<a class="anchor" id="import2">

## 1.2. Import functions

</a>


<a class="anchor" id="import3">

## 1.3. Import dataset

</a>


In [3]:
quiz_iot_events=pd.read_csv('/Users/maceli/Downloads/quiz_iot_events.csv')
quiz_iot_events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69761 entries, 0 to 69760
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   UnitNumber_masked  69761 non-null  object 
 1   start_time         69761 non-null  object 
 2   end_time           69761 non-null  object 
 3   matchcode          69761 non-null  object 
 4   container_status   69761 non-null  object 
 5   inland_move_type   69761 non-null  object 
 6   DP_VOYAGE_NO       69761 non-null  int64  
 7   TIME               69761 non-null  object 
 8   General_Pos_Lat    69761 non-null  float64
 9   General_Pos_Long   69761 non-null  float64
 10  type_of_movement   69761 non-null  object 
 11  eq_type_group      69761 non-null  object 
dtypes: float64(2), int64(1), object(9)
memory usage: 6.4+ MB


In [4]:
quiz_voyages=pd.read_csv('/Users/maceli/Downloads/quiz_voyages.csv')
quiz_voyages.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39 entries, 0 to 38
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   DP_VOYAGE_NO        39 non-null     int64 
 1   MOT_SERVICE         39 non-null     object
 2   SCHEDULE_VOYAGE_NO  39 non-null     object
dtypes: int64(1), object(2)
memory usage: 1.0+ KB


<hr>
<a class="anchor" id="exploration">
    
# 2. Data Exploration
    
</a>

<a class="anchor" id="missing">

## 2.1. Summary statistics and missing values

</a>

In [5]:
summary_statistics(quiz_iot_events)

count unique                               top   freq  \
UnitNumber_masked    69761   1000  57be74833e7aa2101c891a759a0739f9   1159   
start_time           69761   1037        2024-02-06 21:42:00.000000   1159   
end_time             69761   1031        2024-02-21 04:28:00.000000   1159   
matchcode            69761      1                         WASHIN125  69761   
container_status     69761      2                              Full  57688   
inland_move_type     69761      2                            import  57688   
DP_VOYAGE_NO       69761.0    NaN                               NaN    NaN   
TIME                 69761  68866        2024-02-04 19:47:21.000000      3   
General_Pos_Lat    69761.0    NaN                               NaN    NaN   
General_Pos_Long   69761.0    NaN                               NaN    NaN   
type_of_movement     69761      6                              stay  32763   
eq_type_group        69761      5                              45GP  42894   

                            mean         std        min        25%        50%  \
UnitNumber_masked            NaN         NaN        NaN        NaN        NaN   
start_time                   NaN         NaN        NaN        NaN        NaN   
end_time                     NaN         NaN        NaN        NaN        NaN   
matchcode                    NaN         NaN        NaN        NaN        NaN   
container_status             NaN         NaN        NaN        NaN        NaN   
inland_move_type             NaN         NaN        NaN        NaN        NaN   
DP_VOYAGE_NO       626662.746979  1195.15767   624533.0   626984.0   627192.0   
TIME                         NaN         NaN        NaN        NaN        NaN   
General_Pos_Lat        44.808213    3.422008        0.0    41.7534   47.12049   
General_Pos_Long     -112.990457   13.606868 -130.36133 -122.38954 -122.23643   
type_of_movement             NaN         NaN        NaN        NaN        NaN   
eq_type_group                NaN         NaN        NaN        NaN        NaN   

                         75%       max  
UnitNumber_masked        NaN       NaN  
start_time               NaN       NaN  
end_time                 NaN       NaN  
matchcode                NaN       NaN  
container_status         NaN       NaN  
inland_move_type         NaN       NaN  
DP_VOYAGE_NO        627256.0  628087.0  
TIME                     NaN       NaN  
General_Pos_Lat     47.25965  54.28601  
General_Pos_Long  -101.14894       0.0  
type_of_movement         NaN       NaN  
eq_type_group            NaN       NaN

In [6]:
missing_data(quiz_iot_events)

Total  Percent
UnitNumber_masked      0      0.0
start_time             0      0.0
end_time               0      0.0
matchcode              0      0.0
container_status       0      0.0
inland_move_type       0      0.0
DP_VOYAGE_NO           0      0.0
TIME                   0      0.0
General_Pos_Lat        0      0.0
General_Pos_Long       0      0.0
type_of_movement       0      0.0
eq_type_group          0      0.0

In [7]:
summary_statistics(quiz_voyages)

count unique   top freq           mean         std  \
DP_VOYAGE_NO        39.0    NaN   NaN  NaN  626897.846154  987.189193   
MOT_SERVICE           39      7   PN1   15            NaN         NaN   
SCHEDULE_VOYAGE_NO    39     35  008E    2            NaN         NaN   

                         min       25%       50%       75%       max  
DP_VOYAGE_NO        624533.0  626997.0  627195.0  627256.5  628087.0  
MOT_SERVICE              NaN       NaN       NaN       NaN       NaN  
SCHEDULE_VOYAGE_NO       NaN       NaN       NaN       NaN       NaN

In [8]:
missing_data(quiz_voyages)

Total  Percent
DP_VOYAGE_NO            0      0.0
MOT_SERVICE             0      0.0
SCHEDULE_VOYAGE_NO      0      0.0

In [9]:
print(quiz_iot_events['TIME'].min())
print(quiz_iot_events['TIME'].max())

print(quiz_iot_events['start_time'].min())
print(quiz_iot_events['start_time'].max())

print(quiz_iot_events['end_time'].min())
print(quiz_iot_events['end_time'].max())

2024-01-23 00:29:49.000000
2024-03-31 01:55:47.000000
2024-01-23 00:06:00.000000
2024-03-28 16:10:00.000000
2024-01-24 22:32:00.000000
2024-03-31 02:19:00.000000


<hr>
<a class="anchor" id="questions">
    
# 3. Questions
    
</a>

<a class="anchor" id="q1">

## 3.1. How many containers are exported and imported from the port?


</a>

In [10]:
keep_col=['UnitNumber_masked']
df_dp=quiz_iot_events[quiz_iot_events.columns[quiz_iot_events.columns.isin(keep_col)]]
df_dp.drop_duplicates()


UnitNumber_masked
0      5e4a82473c934a548675b920d72ca241
44     cf8041f1de8b43f17d17cb0c3b938862
74     fdf2bedc6bac0817572b020591753afd
91     3fdbd437d300ed3a458c5f922365eb7c
151    13bcc3293aceb2fede32554b132afd5b
...                                 ...
69241  bdcdebca7da92675f9fa678b66007dd4
69287  e0c745bb17cf544ff9fdd4fa774b4f32
69288  e98dc90850ccd8dc9c59e4f158eb687d
69327  7fe1e3f057a0fd9e2aba7ab1d957bdf9
69543  644d9e111fcd1a3b064d98b31e447479

[1000 rows x 1 columns]

In this case UnitNumber_masked is primary key. Each UnitNumber_masked represents one container.

In [11]:
quiz_iot_events['UnitNumber_masked'].nunique()
print(f"Containers exported and imported from the port: {(quiz_iot_events['UnitNumber_masked'].nunique())}")


Containers exported and imported from the port: 1000


<a class="anchor" id="q2">

## 3.2 How many containers are exported and imported per shipsystem (SSY)?                    

SSYs are the different ocean services which are connected to voyages. In the data set, the more
generalized term MOT_Service is used (MOT = Mode of Transport)

</a>

In [12]:
quiz_iot_events['DP_VOYAGE_NO'].nunique()

36

In [13]:
quiz_voyages['DP_VOYAGE_NO'].nunique()

39

In [14]:
col_filt=['UnitNumber_masked','DP_VOYAGE_NO']
quiz_iot_k=quiz_iot_events[quiz_iot_events.columns[quiz_iot_events.columns.isin(col_filt)]]
quiz_iot_k.drop_duplicates (subset=['UnitNumber_masked','DP_VOYAGE_NO'], inplace=True)  
quiz_iot_k.reset_index(drop=True, inplace=True)
quiz_iot_k.head(10)

<ipython-input-14-c0dabdf9aa86>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  quiz_iot_k.drop_duplicates (subset=['UnitNumber_masked','DP_VOYAGE_NO'], inplace=True)


UnitNumber_masked  DP_VOYAGE_NO
0  5e4a82473c934a548675b920d72ca241        627195
1  cf8041f1de8b43f17d17cb0c3b938862        624559
2  fdf2bedc6bac0817572b020591753afd        627190
3  3fdbd437d300ed3a458c5f922365eb7c        627258
4  13bcc3293aceb2fede32554b132afd5b        624556
5  615c58f79773b4dfb69ebcaf46842de8        627204
6  6b0636837ade9b1720710a49d9d20fae        624556
7  3d0b215eb46e2a1c8384e6dfe512b0a0        627195
8  3d0b215eb46e2a1c8384e6dfe512b0a0        627204
9  96d0082ecf9e44638d57959c18d2b79f        624556

In [15]:
quiz_iot_voyage=quiz_iot_k.merge(quiz_voyages, on='DP_VOYAGE_NO',how='left')

In [16]:
quiz_iot_voyage_g=quiz_iot_voyage.groupby("MOT_SERVICE")['UnitNumber_masked'].count()
quiz_iot_voyage.shape

(1067, 4)

In [17]:
print(f"Containers distribution per shipsystem (SSY): {(quiz_iot_voyage_g)}")


Containers distribution per shipsystem (SSY): MOT_SERVICE
AL7      2
CES      1
PN1    357
PN2      3
PN4    703
WSN      1
Name: UnitNumber_masked, dtype: int64


<a class="anchor" id="q3">

## 3.3 Display the IoT on a map by highlighting the points and by displaying the routes per move?

</a>

To plot the routes type_of_movement: 'stay' and 'moving' will removed

In [18]:
sample_df = quiz_iot_events.sample(n=20)
sample_f=sample_df['UnitNumber_masked'].unique()

In [19]:
df_iot_f=quiz_iot_events[(~quiz_iot_events['type_of_movement'].isin(['stay','moving']))
& (quiz_iot_events['UnitNumber_masked'].isin(sample_f))]
df_iot_f['UnitNumber_masked'].nunique()

20

In [20]:
q3 = folium.Map()
routes_map(df_iot_f,q3)

<a class="anchor" id="q4">

## 3.4 What is the maximum distance a container traveled from the hinterland to a port (and vice versa)??

</a>

In [21]:
quiz_iot_events['Distance'] = np.where(quiz_iot_events['UnitNumber_masked']==quiz_iot_events['UnitNumber_masked'].shift(),
                         haversine(quiz_iot_events['General_Pos_Lat'].shift(), 
                                   quiz_iot_events['General_Pos_Long'].shift(), 
                                   quiz_iot_events['General_Pos_Lat'], quiz_iot_events['General_Pos_Long']
                                  ),0)


In [22]:
quiz_iot_events['Distance'].max()
print(f"Maximum distance without stop: {quiz_iot_events['Distance'].max()}")

Maximum distance without stop: 2811.976161317974


In [23]:

quiz_iot_events[quiz_iot_events['Distance']==2811.976161317974]
val_df=quiz_iot_events[quiz_iot_events['UnitNumber_masked']=='70e71af3efb0137fe7769d73ba460072']

In [24]:
q3_val = folium.Map()
routes_map(val_df,q3_val)

In [25]:
df_sum=quiz_iot_events.groupby("UnitNumber_masked")['Distance'].sum().reset_index()
df_sum['Distance'].max()

8833.813512451321

In [26]:
df_sum[df_sum['Distance']==8833.813512451317]
val_df2=quiz_iot_events[(quiz_iot_events['UnitNumber_masked']=='2fe646a92d58075097c8c27dd9da50fe')
&(~quiz_iot_events['type_of_movement'].isin(['stay','moving']))
].reset_index()

val_df2['Distance'] = np.where(val_df2['UnitNumber_masked']==val_df2['UnitNumber_masked'].shift(),
                         haversine(val_df2['General_Pos_Lat'].shift(), 
                                   val_df2['General_Pos_Long'].shift(), 
                                   val_df2['General_Pos_Lat'], val_df2['General_Pos_Long']
                                  ),0)
val_df2=val_df2[val_df2['type_of_movement']=='entering pos']

In [27]:
print(f"Maximum distance considering all stops: {df_sum['Distance'].max()}")
print(f"Maximum distance not considering stay and moving: {val_df2['Distance'].sum()}")

Maximum distance considering all stops: 8833.813512451321
Maximum distance not considering stay and moving: 7613.853084733668


In [28]:
#Ploting considering only 'type_of_movement']=='entering pos'
map_val2 = folium.Map()
routes_map(val_df2,map_val2)

<a class="anchor" id="q5">

## 3.5 Calculate the travel time of the containers and give an insight on the typical travel times ofthe containers.

</a>

In [29]:
quiz_iot_events=date_diff(quiz_iot_events,'start_time','end_time')

In [30]:
quiz_iot_events['travel_time'].describe()

count                         69761
mean     14 days 09:06:34.510686486
std       6 days 16:51:11.677697995
min                 1 days 04:10:00
25%                 9 days 11:20:00
50%                13 days 12:28:00
75%                17 days 17:09:00
max                43 days 10:46:00
Name: travel_time, dtype: object

In [31]:
cv=quiz_iot_events['travel_time'].std()/quiz_iot_events['travel_time'].mean()*100
cv

46.60932358579309

With cv 46% indicates that there is a relatively high level of variability in the travel time of containers compared to the mean. As we can see from the table, the travel time range is between 1 day to 43 days.

<a class="anchor" id="q6">

## 3.6 General: Share your observations on the data?.

</a>

1. Date quality: not a problem
2. Even with 1000 distincit UnitNumber_masked, was possible get insights
3. ~65% of containers are distributed by PN4
4. 75% of the container has a travel time under 18 days.